src/main/java/io/gravitee/policy/apikey/ApiKeyPolicy.java

```diff
@@ -176,11 +182,103 @@ private Optional<String> extractApiKey(HttpExecutionContext ctx) {
         return Optional.ofNullable(apiKey);
     }
 
-    private void cleanupApiKey(HttpExecutionContext ctx) {
+    private void cleanupApiKey(HttpPlainExecutionContext ctx) {
         if (!propagateApiKey) {
             ctx.request().headers().remove(API_KEY_HEADER);
             ctx.request().parameters().remove(API_KEY_QUERY_PARAMETER);
         }
         ctx.removeInternalAttribute(ATTR_INTERNAL_API_KEY);
     }
+
+    @Override
+    public Maybe<SecurityToken> extractSecurityToken(KafkaConnectionContext ctx) {
+        Callback[] callbacks = ctx.callbacks();
+        for (Callback callback : callbacks) {
+            if (callback instanceof NameCallback nameCallback) {
+                // With SASL_PLAIN or SCRAM, we expect the username to be a md5 hash of the api-key, for security and privacy.
+                String md5ApiKey = nameCallback.getName();
+                if (md5ApiKey != null && !md5ApiKey.isBlank()) {
+                    ctx.setInternalAttribute(ATTR_INTERNAL_MD5_API_KEY, md5ApiKey);
+                    return Maybe.just(SecurityToken.forMD5ApiKey(md5ApiKey));
+                }
+                return Maybe.just(SecurityToken.invalid(MD5_API_KEY));
+            }
+        }
+        return Maybe.empty();
+    }
+
+    @Override
+    public Completable authenticate(KafkaConnectionContext ctx) {
```

Comment:
@jhaeyaert should it be executed in a Completable.defer() ?

Reply:
Yes 👍